### ISSUE TITLE
Multi threading phase 2: parallel serialization of character columns

### ISSUE THREAD

**Issue by fstpackage** [fstpackage/fst#63]
Parallelization of character columns requires more work for the `R` framework than it's integer, double and logical counterparts. That's because `R` uses a global string pool and each string must be allocated on the master thread with methods from the `R` API. That API is relatively slow and the reason for sub-optimal performance of character vector serialization. The fastest method for storing a character vector is probably:

1) factorize the character vector from C++ using the individual string element pointer values. Because of `R`'s string pool, identical strings in a character vector have an identical pointer address to the underlying value (const char* in C++). In effect, the vector is already factorized. We can use that fact to do a very fast factorization of the character vector. This can also be done in parallel (with limitations).
2) Serialize the factor: in general, the level vector (a character vector) will be smaller than the underlying integer vector and the latter can be serialized effectively using multiple threads.
3) Upon de-serialization, we can retrieve the factor or use the level information to copy identical strings (faster than allocating through the string pool)

**Comment by MarcusKlik** [comment]
The slowest part of de-serializing a `character` vector is asking `R`'s global string pool to create a new `string` or return an already existing one, see for example [this code](https://github.com/fstpackage/fst/blob/develop/src/fst_blockrunner_char.cpp#L153).
The `R` API to the global string pool can be used from the main thread only, so we can't create new strings from multiple threads simultaneously like with the other column types. When each element from a character vector is created with a call to `Rf_mkCharLenCE` (or equivalent), the speed will never get above a GB/s (probably only half of that).

So the only way to beat `R`'s slow global string pool is do the string lookup from `C++` and use the `R` API only to copy already existing elements from a vector to new positions.

For example, when a 1 million `character` vector has 100 unique values, we only need to call `Rf_mkCharLenCE` for 100 elements. For the other 999900 elements, we just copy the appropriate, already existing element, to a new location. That way, we avoid expensive (additional) look-ups in the global string pool,

But `fst` has full random access. That means that we need to implement this technique for each data block (2048 string elements) in a vector. Any duplicates in that block can be copied instead of created through the `R` API. Possible candidates to determine unique strings in those (size 2048) string blocks are:

* `std::unordered_map` (standard `C++`, so easiest)
* `boost::unordered_map` (requires adding a lot of code but faster)
* custom unordered hash map for pointer values (small code addition and fastest solution)

Tests are required to test how fast these maps can find unique values in a (at maximum) 2048 element pointer vector. To be effective, the single threaded speed should be above 150 MB/s and the algorithm should scale well with additional cores. A map object that only exists in local cache memory would be ideal and would probably scale well with multiple threads.

**Comment by MarcusKlik** [comment]
Preliminary results show that a custom made hash map can get to speeds faster than 200 MB/s per core. Below are the results of creating a hash map of a length 1e6 `character` vector. The mapping is done in blocks of 2048 elements each and the average size of a `character` element is 4.

![image](https://user-images.githubusercontent.com/24658401/36874333-d3ec9466-1dab-11e8-99bb-13a749517b20.png)

The figure compares the speed of the `C++` `std::unordered_map`, `boost::unordered_map` and a custom made stack based hash map (`fstlib::unordered_stack_map`). The standard `C++` `unordered_map` is far too slow to be used for factorization of a 2048 elements `character` vector in `fst`. The same holds for the `boost` implementation, although it is significantly faster than `std`. For real-time factorization, we need speeds higher than 150 MB per core, even for a larger number of unique values.

The best performance was measured with `fstlib::unordered_stack_map` using 4096 buckets (to sort 2048 elements). Each bucket contains a linked list of pointer addresses and values. The custom implementation is fully stack based for speed and uses somewhat less than 64 kB of cache memory. I hope that won't introduce any problems on older CPU's (with smaller L1 - L3 cache sizes). Note that the measured speed will scale linearly with the average element size. So for an average string size of 8, speeds will double.

Now that we have an `unordered_map` with appropriate speed, `character` columns can be serialized with the following technique:

* The `character` column is processed in blocks of 2048 elements (strings) each.
* We read the pointer addresses of all `string` elements in the block.
* The `fstlib::unordered_stack_map` is used to determine the unique addresses in the block. The amount of unique addresses will be equal to the number of unique `strings` in the block. That's because `R` has a global string pool and identical `strings` point to identical memory addresses.
* The unique `strings` are stored as before along with an index that connects each element to the correct (unique) `string` value. For a small amount of unique values, this will significantly reduce the size of the stored data.

Upon reading the column, we perform the following steps:

* Reading is done in block of 2048 elements.
* For each block, the unique values are read along with the (length 2048) index.
* The `R` API is used to add each unique `string` in the block to `R`'s string pool. Because we already know the unique values, the API will be called (far) fewer times.
* For the other elements, we just copy the allocated unique values. This circumvents the string pool allocation.

Using this scheme, `R`'s slow string pool won't be a bottleneck anymore for reading `character` columns, as the mapping is already done in `C++` and `R` has to map far fewer elements (depending on the number of unique values off course). Moreover, the current size of `R`'s string pool won't determine the lookup speed as much anymore. In the current setup, if a string pool lookup is done for each element in a vector and the string pool is already large, the speed suffers significantly. For writing, less copies from the main memory are required, so that might increase speed as well.

With a few hundred MB/s per core, the multi-core speed of this scheme is expected to be scalable to multiple GB's per second for columns that have large duplication of elements (where the total number of elements is larger than the number of unique values).
